kernel/src/hil/adc.rs

```diff
@@ -1,38 +1,80 @@
+/// Interfaces for analog to digital converter peripherals.
 use returncode::ReturnCode;
 
-/// Trait for handling callbacks from ADC module.
-pub trait Client {
-    /// Called when a sample is ready.
-    fn sample_done(&self, sample: u16);
-}
 
-/// Simple interface for reading a single ADC sample on any channel.
-pub trait AdcSingle {
+// *** Interfaces for low-speed, single-sample ADCs ***
+
+/// Simple interface for reading an ADC sample on any channel.
+pub trait Adc {
+    /// The chip-dependent type of an ADC channel.
+    type Channel;
+
     /// Initialize must be called before taking a sample.
-    /// Returns true on success.
     fn initialize(&self) -> ReturnCode;
 
     /// Request a single ADC sample on a particular channel.
-    /// Returns true on success.
-    fn sample(&self, channel: u8) -> ReturnCode;
-    fn cancel_sample(&self) -> ReturnCode;
+    /// Used for individual samples that have no timing requirements.
+    fn sample(&self, channel: &Self::Channel) -> ReturnCode;
+
+    /// Request repeated ADC samples on a particular channel.
+    /// Callbacks will occur at the given frequency with low jitter and can be set to any frequency
+    /// supported by the chip implementation. However callbacks may be limited based on how quickly
+    /// the system can service individual samples, leading to missed samples at high frequencies.
+    fn sample_continuous(&self, channel: &Self::Channel, frequency: u32) -> ReturnCode;
+
+    /// Stop a sampling operation.
+    /// Can be used to stop simple or high-speed samples. No further callback will occur in simple
+    /// mode. The `sampling_complete` callback will occur in high-speed mode.
+    fn stop_sampling(&self) -> ReturnCode;
```

Comment:
After a night to sleep on it, I'm a little less satisfied with changing the `stop_sampling` interface since it would have to return `(ReturnCode, None, None)` for the normal sampling case (since the same `stop_sampling` command applied to both normal and highspeed mode).